File: backend/app/schemas/enrichment.py
"""Pydantic schemas for Video Enrichment API endpoints."""

from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, Field


class EnrichmentStatus(str, Enum):
    """Status of video enrichment processing."""

    pending = "pending"
    processing = "processing"
    completed = "completed"
    partial = "partial"
    failed = "failed"


class ChapterSchema(BaseModel):
    """Schema for a video chapter."""

    title: str = Field(..., description="Chapter title")
    start: float = Field(..., description="Start time in seconds")
    end: float = Field(..., description="End time in seconds")


class EnrichmentResponse(BaseModel):
    """Response schema for video enrichment data."""

    id: UUID = Field(..., description="Enrichment record ID")
    video_id: UUID = Field(..., description="Associated video ID")
    status: EnrichmentStatus = Field(..., description="Enrichment processing status")

    # Captions
    captions_vtt: str | None = Field(None, description="Captions in VTT format")
    captions_language: str | None = Field(None, description="Caption language code")
    captions_source: str | None = Field(
        None, description="Caption source (youtube_manual, youtube_auto, groq_whisper)"
    )

    # Transcript
    transcript_text: str | None = Field(
        None, description="Plain text transcript for search"
    )

    # Chapters
    chapters: list[ChapterSchema] | None = Field(None, description="Video chapters")
    chapters_vtt: str | None = Field(None, description="Chapters in VTT format")
    chapters_source: str | None = Field(
        None, description="Chapter source (youtube, description)"
    )

    # Thumbnails
    thumbnails_vtt_url: str | None = Field(
        None, description="Thumbnails sprite sheet VTT URL"
    )

    # Processing state
    error_message: str | None = Field(None, description="Error message if failed")
    retry_count: int = Field(0, description="Number of retry attempts")
    progress_message: str | None = Field(None, description="Current processing step")

    # Timestamps
    created_at: datetime | None = Field(None, description="Record creation time")
    updated_at: datetime | None = Field(None, description="Last update time")
    processed_at: datetime | None = Field(None, description="Completion time")

    model_config = {"from_attributes": True}


class EnrichmentRetryResponse(BaseModel):
    """Response schema for enrichment retry request."""

    message: str = Field(..., description="Status message")
    enrichment: EnrichmentResponse = Field(..., description="Updated enrichment data")

File: frontend/src/types/enrichment.ts
import { z } from "zod";

// ============================================================================
// Enrichment Status
// ============================================================================

/**
 * Status of video enrichment processing.
 * Matches backend EnrichmentStatus enum.
 */
export const EnrichmentStatusSchema = z.enum([
  "pending",
  "processing",
  "completed",
  "partial",
  "failed",
]);

export type EnrichmentStatus = z.infer<typeof EnrichmentStatusSchema>;

// ============================================================================
// Chapter Types
// ============================================================================

/**
 * Zod schema for a video chapter.
 * Represents a titled segment of a video with start/end times.
 */
export const ChapterSchema = z.object({
  title: z.string(),
  start: z.number(), // Start time in seconds
  end: z.number(), // End time in seconds
});

export type Chapter = z.infer<typeof ChapterSchema>;

// ============================================================================
// Enrichment Response
// ============================================================================

/**
 * Zod schema for EnrichmentResponse.
 * Matches backend app/schemas/enrichment.py EnrichmentResponse.
 *
 * Contains captions, chapters, thumbnails, and processing state.
 */
export const EnrichmentResponseSchema = z.object({
  id: z.string().uuid(),
  video_id: z.string().uuid(),
  status: EnrichmentStatusSchema,

  // Captions
  captions_vtt: z.string().nullable().optional(),
  captions_language: z.string().nullable().optional(),
  captions_source: z.string().nullable().optional(), // 'youtube_manual', 'youtube_auto', 'groq_whisper'

  // Transcript
  transcript_text: z.string().nullable().optional(),

  // Chapters
  chapters: z.array(ChapterSchema).nullable().optional(),
  chapters_vtt: z.string().nullable().optional(),
  chapters_source: z.string().nullable().optional(), // 'youtube', 'description'

  // Thumbnails
  thumbnails_vtt_url: z.string().nullable().optional(),

  // Processing state
  error_message: z.string().nullable().optional(),
  retry_count: z.number().default(0),
  progress_message: z.string().nullable().optional(),

  // Timestamps
  created_at: z.string().nullable().optional(),
  updated_at: z.string().nullable().optional(),
  processed_at: z.string().nullable().optional(),
});

export type EnrichmentResponse = z.infer<typeof EnrichmentResponseSchema>;

// ============================================================================
// Enrichment Retry Response
// ============================================================================

/**
 * Response from POST /api/videos/{id}/enrichment/retry
 */
export const EnrichmentRetryResponseSchema = z.object({
  message: z.string(),
  enrichment: EnrichmentResponseSchema,
});

export type EnrichmentRetryResponse = z.infer<
  typeof EnrichmentRetryResponseSchema
>;

// ============================================================================
// Helper Types
// ============================================================================

/**
 * Caption source labels for display
 */
export const CAPTION_SOURCE_LABELS: Record<string, string> = {
  youtube_manual: "YouTube (Manual)",
  youtube_auto: "YouTube (Auto)",
  groq_whisper: "Groq Whisper",
};

/**
 * Chapter source labels for display
 */
export const CHAPTER_SOURCE_LABELS: Record<string, string> = {
  youtube: "YouTube",
  description: "Description",
};

/**
 * Language code to human-readable name mapping
 */
export const LANGUAGE_LABELS: Record<string, string> = {
  en: "English",
  de: "German",
  es: "Spanish",
  fr: "French",
  it: "Italian",
  pt: "Portuguese",
  ja: "Japanese",
  ko: "Korean",
  zh: "Chinese",
  ru: "Russian",
};
